refactor(protocol-engine): Implement `moveToCoordinates`

SyntaxColoring opened the pull request:
# Overview

This PR implements the Protocol Engine `moveToCoordinates` command. In the short term, this is necessary to fully implement the JSON Protocol v6 spec. In the longer term, it will be important in the implementation of Python Protocol API v3.

Closes #10757. See that ticket for the API specification.

# Changelog

* Implement the `moveToCoordinates` Protocol Engine command as previously specified.
* Fix a `float`->`int` truncation bug in the `JSONCommandTranslator`.
* Refactor the tests for `MotionView.get_waypoints_to_well()`.

# Manual testing

I think we currently have to test this by running commands on a real robot. And we have to do it over HTTP, since PAPIv3 isn't implemented enough to emit this command, and I assume Protocol Designer doesn't currently emit this command.

I've hacked together some scripts to help with this in Opentrons/otietalk#1.

Some interesting things to look out for:

* [x] When the pipette has a tip, the bottom of the *tip* should move to the given coordinates. When it doesn't, the bottom of the *nozzle* should move to the given coordinates.
* [x] When an 8-Channel pipette is used, the *back-most* channel should move to the given coordinates.
* [x] By default, the pipette should automatically move high enough to clear all loaded labware and modules.
* [x] In the following sequence, the pipette should arc for both movements. It should not move directly.

     1. Move to a well.
     2. `moveToCoordinates`.
     3. Move back to the original well.

* [ ] The handling of critical points is a little delicate and I only tenuously understand it.

    One case I know of where this matters is when an 8-Channel pipette is used, and the pipette starts off in a labware with the `centerMultichannelOnWells` quirk, like a reservoir or trough. Supposing the pipette got there with a `moveToWell` or similar, its initial "logical location" will be at the *center* of the well—even though the back-most channel is somewhere behind the center. If I got something wrong with critical points, the initial "move up" step of the arc will wrongly include an unexpected XY motion in addition to its normal Z motion.

# Other review requests

* [ ] Is the handling of critical points correct?
* [ ] Is the handling of critical points understandable? Can I explain it better through better comments, for example?
* [ ] As implemented, this will not dodge the Thermocycler like `moveToWell` does. This matches PAPIv2's behavior. We agree that in the future, we want more sophisticated collision checking based on geometry and coordinates and bounding boxes, rather than logical wells. **But for now:** Do we want to document anywhere that `moveToCoordinates` does not dodge the Thermocycler? Note that `moveToWell` does not currently document that it *does* dodge the Thermocycler.
* [ ] Is there any other location-tracking code I missed that needs to be adjusted?

# Risk assessment

Medium.

This isn't used by anything today as far as I know, so there's no risk to changing production behavior right now. But, it will be used in the future, and it is new motion planning code, which is sensitive. See the testing notes and review requests.

SyntaxColoring committed:
Rename `get_movement_waypoints()` `get_movement_waypoints_to_well()`.

SyntaxColoring committed:
Leave todo for raising error on out-of-bounds direct movements.

SyntaxColoring committed:
WIP: Roughly implement moving to coordinates.

SyntaxColoring committed:
Move implementation to command file and implement minimumZHeight.

SyntaxColoring committed:
Update old note about experimental waypoint planning.

SyntaxColoring committed:
Rename `module_view` -> `mock_module_view` in tests.

SyntaxColoring committed:
Convert `get_waypoints()` usage in tests to mock.

SyntaxColoring committed:
Fix typing error.

SyntaxColoring committed:
Format.

SyntaxColoring committed:
Convert `get_movement_waypoints_to_well()`'s raise test to mock, too.

SyntaxColoring committed:
Update todos for big `test_get_movement_waypoints_to_well()`.

SyntaxColoring committed:
Make min_travel_z non-optional.

SyntaxColoring committed:
Add unit tests for `get_movement_waypoints_to_coords()`.

SyntaxColoring commented on api/src/opentrons/motion_planning/waypoints.py:
```diff
@@ -18,7 +18,7 @@ def get_waypoints(
     *,
     max_travel_z: float,
-    min_travel_z: float = 0.0,
+    min_travel_z: float,
```
I made this argument mandatory because:

* Only one place was not providing an explicit value.
* I think we want to encourage callers to think about this value. It's likely unsafe to leave this at `0` because this needs to be high enough to clear any labware on the deck.

SyntaxColoring replied:
I made this argument mandatory because:

* Only one place was not providing an explicit value.
* I think we want to encourage callers to think about this value. It's likely unsafe to leave this at `0` because this needs to be high enough to clear any labware on the deck.

SyntaxColoring committed:
Format and lint.

SyntaxColoring committed:
Test `moveToCoordinates` implementation.

SyntaxColoring committed:
Add JSONCommandTranslator test and fix type truncation bug.

SyntaxColoring committed:
Use a less intrusive workaround for the () vs. [] arg matching problem.

SyntaxColoring commented on api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py:
```diff
@@ -11,15 +15,84 @@
+        state_view.motion.get_movement_waypoints_to_coords(
+            origin=current_position,
+            dest=Point(
+                params.coordinates.x, params.coordinates.y, params.coordinates.z
```
I keep going back and forth on whether it's a good idea for unit tests to reuse common values like this, versus copy-pasting a bunch of magic inline values like `Point(x=1.11, y=2.22, z=3.33)`.

On one hand, if every step redefines its own literals like `Point(x=1.11, y=2.21, z=3.13)`, things can get hard to read in longer tests like this one.

On the other hand, if we use common variables like this, the test is basically just a reimplementation of the production code.

mcous replied:
When in doubt, I try to use value literals in test expectations

SyntaxColoring replied:
How strongly do you feel about that, in this case?

[Here's how this test looks with literals](https://gist.github.com/SyntaxColoring/89078af6292e4179cebcfbd613ee553a). I personally find it pretty hard to follow the dataflow. Maybe this method has an unusually high number of values that have the same type? For example, there are at least 4 different `Point`s that ought to be thought of separately.

mcous replied:
Not super strongly, just a rule of thumb to start from. If you find it easier to follow with the variables, then I think my only suggestion would be to name them `expected_whatever` whenever possible to make it clear that it's a variable we invented for the test

SyntaxColoring commented on api/tests/opentrons/protocol_engine/state/test_motion_view.py:
```diff
@@ -198,19 +217,18 @@ class WaypointSpec:
     all_labware_z: Optional[float] = None
     max_travel_z: float = 50
     should_dodge_thermocycler: bool = False
-    extra_waypoints: Sequence[Tuple[float, float]] = ()
+    extra_waypoints: Sequence[Tuple[float, float]] = field(default_factory=list)
```
This change was necessary to work around a silly argument matching thing. The production code passes `[]` when there are no extra waypoints, not `()`. This discrepancy started to matter now that the tests are written to use `get_waypoints()` as a Decoy mock.

mcous replied:
Oof. Hopefully `WaypointSpec` can go away in favor of the orthogonal tests you mention below

SyntaxColoring commented on api/tests/opentrons/protocol_engine/state/test_motion_view.py:
```diff
@@ -198,19 +217,18 @@ class WaypointSpec:
     all_labware_z: Optional[float] = None
     max_travel_z: float = 50
     should_dodge_thermocycler: bool = False
-    extra_waypoints: Sequence[Tuple[float, float]] = ()
+    extra_waypoints: Sequence[Tuple[float, float]] = field(default_factory=list)
 
 
-# TODO(mc, 2021-01-14): these tests probably need to be rethought; this fixture
-# is impossible to reason with. The test is really just trying to be a collaborator
-# test for the `get_waypoints` function, so we should rewrite it as such.
+# TODO(mm, 2022-06-22): Break this test apart into the parts that are orthogonal
+# (see this test's docstring) or refactor the subject to do less at once.
```
Per discussion with @mcous, I've reworked this test a little bit.

Formerly, this test called the raw `get_waypoints()` function how it expected the *subject* to call it. Then, it compared the results from the raw call to the results from the subject.

Now, it sets up `get_waypoints()` as a monkey-patched mock. It can now assert that the subject calls `get_waypoints()` correctly through our usual Decoy mocking techniques.

I think this test is still hard to follow, but further refactors turned out to be a bit much for this PR, so I left it alone other than that change.

SyntaxColoring commented on shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py:
```diff
@@ -43,7 +43,7 @@ class Params(BaseModel):
     wellLocation: Optional[WellLocation]
     wait: Optional[int]
-    minimumZHeight: Optional[int]
+    minimumZHeight: Optional[float]
```
To match the JSON protocol schema:

https://github.com/Opentrons/opentrons/blob/da09f5f490d51c79f4398b54164b542cfede78a8/shared-data/protocol/schemas/6.json#L420-L424

And our main Python model:

https://github.com/Opentrons/opentrons/blob/da09f5f490d51c79f4398b54164b542cfede78a8/api/src/opentrons/protocol_engine/commands/move_to_coordinates.py#L30

SyntaxColoring commented on api/src/opentrons/protocol_engine/state/motion.py:
```diff
@@ -134,5 +128,50 @@ def get_movement_waypoints(
+                Ignored if `direct` is True. If lower than the default height,
+                the default is used; this can only increase the height, not decrease it.
+        """
+        all_labware_highest_z = self._geometry.get_all_labware_highest_z()
```
This reliance on `get_all_labware_highest_z()` matches how `get_movement_waypoints_to_well()` works. But looking at this again, I think this might be a bug because it only takes labware into account, not modules.

Ticket this separately? Then we can:

* Double-check what PAPIv2 does
* Fix this (if necessary) for movements to coordinates and movements to wells
* Test it on a real robot with real modules

mcous replied:
Oh that's not good. PAPIv2 certainly takes all items into account (via the `Deck` class). Looks like we ignore modules if there's no labware on it. We'll want to resolve this, potentially in the 6.0.0 release. Will investigate today

SyntaxColoring replied:
Addressed in #10902.

SyntaxColoring commented:
Pretty sure I need to update this, too:

https://github.com/Opentrons/opentrons/blob/da09f5f490d51c79f4398b54164b542cfede78a8/api/src/opentrons/protocol_engine/state/pipettes.py#L74-L89

SyntaxColoring committed:
Merge branch 'edge' into pe_move_to_coordinates

Resolve conflicts in:
* shared-data/python/opentrons_shared_data/protocol/models/protocol_schema_v6.py

mcous commented on api/src/opentrons/protocol_engine/commands/move_to_coordinates.py:
```diff
@@ -1,13 +1,20 @@
 from .command import AbstractCommandImpl, BaseCommand, BaseCommandCreate
 
+if TYPE_CHECKING:
+    from opentrons.hardware_control import HardwareControlAPI
```
Does this need to be under the `TYPE_CHECKING` guard?

SyntaxColoring replied:
Nope! Removed.

mcous commented on api/src/opentrons/protocol_engine/commands/move_to_coordinates.py:
```diff
@@ -55,12 +62,70 @@ class MoveToCoordinatesImplementation(
+        direct: bool,
+        additional_min_travel_z: Optional[float],
+    ) -> None:
+        hw_mount = self._state_view.pipettes.get(
+            pipette_id=pipette_id
+        ).mount.to_hw_mount()
```
I'm down for a `get_hardware_mount` selector; I think I've seen this same think in a few places

mcous commented on api/src/opentrons/protocol_engine/state/motion.py:
```diff
@@ -1,15 +1,10 @@
 """Motion state store and getters."""
 from dataclasses import dataclass
-from typing import List, Optional
+from typing import List, Optional, Sequence, Tuple
```
`Sequence` and `Tuple` are unused which is causing linting failures

mcous commented on api/tests/opentrons/protocol_engine/commands/test_move_to_coordinates.py:
```diff
@@ -11,15 +15,84 @@
+
+    result = await subject.execute(params=params)
+
+    decoy.verify(
+        *[
+            await hardware_api.move_to(
+                mount=mount,
+                abs_position=waypoint.position,
+                critical_point=waypoint.critical_point,
+            )
+            for waypoint in planned_waypoints
+        ]
+    )
```
Can we unroll this loop?

mcous commented on api/tests/opentrons/protocol_engine/state/test_motion_view.py:
```diff
@@ -25,24 +25,43 @@
+]
+
+
+@pytest.fixture
+def mock_get_waypoints(
+    decoy: Decoy, monkeypatch: pytest.MonkeyPatch
+) -> _GET_WAYPOINTS_SIGNATURE:
```
Given Python typing does not give us a convenient thing to say "give me the type signature of this function", I think we should:

- Change this to a `None` return and/or `autouse` fixture
- Use `decoy.when(motion_planning.get_waypoints(...))` in tests

Using `_GET_WAYPOINTS_SIGNATURE`, we lose all typechecking on the arguments of `get_waypoints` in tests, which increases the likelihood of miswritten stubs in tests (leading to falsely passing tests) and makes changing the signature of `get_waypoints` harder down the road because we won't have the typechecker guiding us to tests that need to change

SyntaxColoring replied:
Good thinking.

mcous commented on api/tests/opentrons/protocol_engine/state/test_motion_view.py:
```diff
@@ -356,40 +425,161 @@ def test_get_movement_waypoints(
+    decoy.when(
+        # TODO(mm, 2022-06-22): We should use decoy.matchers.Anything() for all
+        # arguments. For some reason, Decoy does not match the call unless we
+        # specify concrete values for all (?) arguments?
```
For the arguments that have default values that you don't care about in this test, you can try using `decoy.when(..., ignore_extra_args=True)`.

- [Usage guide](https://mike.cousins.io/decoy/usage/when/#only-specify-some-arguments)
- [API references](https://mike.cousins.io/decoy/api/#decoy.Decoy.when)

mcous reviewed (commented):
Gonna try testing this out on a robot later today and add another review with testing results. Thanks for investigating `get_all_labware_highest_z`; it looks like a general ProtocolEngine bug to me.

Random thoughts, including some testing recommendations, below

SyntaxColoring committed:
Remove unused imports.

TamarZanzouri reviewed (commented):
Changes looks good to me! Mike gave some pretty good comments (that I would prob miss) and it looks like you fixed some of them already. once CI passes and you are done fixing @mcous comments would love to review this again

mcous reviewed (commented):
Tested on an OT-2R with a P20 Multi GEN2 and the endpoint behaved as expected (barring the fact that #10902 is not yet in `edge` / this branch.

Looks like we'll want to merge in `edge` to grab some CI fixes. From there, we can figure out what we want to do once @SyntaxColoring is back. Seems pretty much ready to go, barring some test tweaks

SyntaxColoring committed:
Merge branch 'edge' into pe_move_to_coordinates

Merging to bring CI fixes and PR #10902 into this branch.

SyntaxColoring commented:
Re-drafting to fix this: https://github.com/Opentrons/opentrons/pull/10840#issuecomment-1164870264

SyntaxColoring committed:
Fix incorrect direct movements in some cases.

When moveToCoordinates was followed by a movement to the same well where the pipette was before the moveToCoordinates, the robot would incorrectly use a direct movement instead of an arc.

SyntaxColoring committed:
Update integration test for newly-preserved `moveToCoordinates` in analysis.

SyntaxColoring committed:
Avoid partially erasing the type of get_waypoints() in tests.

SyntaxColoring committed:
Move HardwareControlAPI out of TYPE_CHECKING guard.

SyntaxColoring committed:
Use ignore_extra_args=True for a bit more conciseness.

SyntaxColoring committed:
Unroll loop.

mcous commented on api/src/opentrons/motion_planning/waypoints.py:
```diff
@@ -41,17 +41,22 @@ def get_waypoints(
     #
-    # Eventually, it may take over for opentrons.hardware_control.util.plan_arc
+    # Flipping `use_experimental_waypoint_planning` to True to make PAPIv2 use this too
+    # causes three test failures at the time of this writing.
```
Just commenting to keep you up to date: I figured out at least two of the test failures! They're caused by behavioral differences in the waypoint planning:

- The old logic ignores `move_type: DIRECT` if you pass a `min_travel_z`, and arcs instead
    - Unclear to me if this behavior is actually used by calling code
    - But there is a test for it! So someone wanted it at some point
- The new logic suppresses duplicate waypoints
    - This behavior is probably unnecessary
    - In recent collaboration sessions, embedded suggested the idea of passing the waypoint list to the HW API rather than doing a loop in calling code. I like this idea, and it might make duplicate waypoint suppression even less important

mcous reviewed (approved):
Good catch on wiping out the last known well
